src/utils/queryOps.js

```diff
@@ -1,225 +1,416 @@
 
-const first = list => {
-  if (list.length === 0) return null;
+  return null;
+}
 
-  return list[0];
-};
+function last(list) {
+  let value = null;
 
-const index = (list, column) => {
-  // Return object indexed on column
-  // We assume the key is unique
-  const output = {};
+  for (const v of list) value = v;
 
-  list.forEach(row => {
-    const key = row[column];
-    const value = output[key];
+  return value;
+}
 
-    if (value === undefined) {
-      // TODO: simplify
-      output[key] = row;
-    } else {
-      throw new Error('expecting index to be unique');
-    }
-  });
+function toArray(value) {
+  // return a list
+  if (Array.isArray(value)) {
+    return value;
+  }
+
+  if (value == null) {
+    return [];
+  }
+
+  return [value];
+}
+
+function zipObject(keys, values) {
+  // accept list of keys and list of values to zip into a single object
+  const output = {};
+
+  for (let i = 0; i < keys.length; i += 1) output[keys[i]] = values[i];
 
   return output;
-};
+}
 
-class Wrapper {
-  constructor(list) {
-    this.list = list;
+function preSelector(columnName) {
+  // convert to an array of [selector(), name] pairs
+  if (Array.isArray(columnName)) {
+    // select many columns
+    return internalFrum(columnName)
+      .map(preSelector)
+      .flatten()
+      .sortBy(([, b]) => b);
   }
 
-  toArray() {
-    return this.list;
+  if (typeof columnName === 'object') {
+    return internalToPairs(columnName)
+      .map((value, name) => [preSelector(value), name])
+      .flatten()
+      .sortBy(([, b]) => b);
   }
 
-  fromPairs() {
-    return lodashFromPairs(this.list);
+  if (typeof columnName === 'string') {
+    return internalFrum([row => row[columnName], columnName]);
   }
+}
 
-  last() {
-    return last(this.list);
-  }
+function selector(columnName) {
+  // convert string into function that selects column from row
+  // convert array of strings into function that extract properties from row
+  // convert an
+  if (typeof columnName === 'object' || Array.isArray(columnName)) {
+    // select many columns
+    const cs = preSelector(columnName).args();
 
-  first() {
-    return first(this.list);
+    return row => cs.map(func => func(row)).fromPairs();
   }
 
-  index(column) {
-    // Return object indexed on column
-    // We assume the key is unique
-    const output = {};
+  if (typeof columnName === 'string') {
+    return row => row[columnName];
+  }
 
-    this.list.forEach(row => {
-      const key = row[column];
-      const value = output[key];
+  return columnName;
+}
 
-      if (value === undefined) {
-        // TODO: simplify
-        output[key] = row;
-      } else {
-        throw new Error('expecting index to be unique');
-      }
-    });
+function missing(value) {
+  // return true if value is null, or undefined, or not a legit value
+  return value == null || Number.isNaN(value) || value === '';
+}
 
-    return output;
-  }
+function exists(value) {
```

Comment:
Could you combine missing and exists?

Reply:
If it was easy to say `!missing`, then yes, but `exists` is defined as `v => !missing(v)`, which requires a bit more thought until you see it as boilerplate.

We could define a `not` functor: `not = f => ((...args) => !f(..args))` and then write `not(missing)`.  But I think this functor would have limited use, and we still have two functions.